parachain/pallets/control/src/lib.rs

```diff
@@ -303,34 +323,59 @@ pub mod pallet {
 	}
 
 	impl<T: Config> Pallet<T> {
-		fn submit_outbound(message: Message) -> DispatchResult {
+		fn submit_outbound(message: Message, agent_location: MultiLocation) -> DispatchResult {
 			let ticket =
 				T::OutboundQueue::validate(&message).map_err(|_| Error::<T>::SubmissionFailed)?;
+			Self::charge_fees(&message, &agent_location)?;
 			T::OutboundQueue::submit(ticket).map_err(|_| Error::<T>::SubmissionFailed)?;
 			Ok(())
 		}
 
 		pub fn convert_location(
 			mut location: MultiLocation,
-		) -> Result<(H256, Option<ParaId>, MultiLocation), DispatchError> {
+		) -> Result<(H256, ParaId, MultiLocation), DispatchError> {
 			// Normalize all locations relative to the relay chain.
 			let relay_location = T::RelayLocation::get();
 			location
 				.reanchor(&relay_location, T::UniversalLocation::get())
 				.map_err(|_| Error::<T>::LocationReanchorFailed)?;
 
-			// Only allow Parachain as origin location
-			let para_id = match location {
-				MultiLocation { parents: 0, interior: X1(Parachain(index)) } =>
-					Some((index).into()),
+			let para_id = match location.interior.first() {
+				Some(Parachain(index)) => Some((*index).into()),
 				_ => None,
-			};
+			}
+			.ok_or(Error::<T>::LocationToParaIdConversionFailed)?;
 
 			// Hash the location to produce an agent id
-			let agent_id = T::AgentHashedDescription::convert_location(&location)
+			let agent_id = T::AgentIdOf::convert_location(&location)
 				.ok_or(Error::<T>::LocationToAgentIdConversionFailed)?;
 
 			Ok((agent_id, para_id, location))
 		}
+
+		pub fn charge_fees(message: &Message, agent_location: &MultiLocation) -> DispatchResult {
+			let agent_account = T::SovereignAccountOf::convert_location(agent_location)
+				.ok_or(Error::<T>::LocationToSovereignAccountConversionFailed)?;
+
+			let fees = T::OutboundQueue::estimate_fee(message)
+				.map_err(|_| Error::<T>::EstimateFeeFailed)?;
+
+			let fee_amount: u128 = match fees.get(0) {
+				Some(&MultiAsset { fun: Fungible(amount), .. }) => amount,
+				_ => 0,
+			};
+
+			if fee_amount > 0 {
+				T::Token::transfer(
+					&agent_account,
+					&T::LocalPalletId::get().into_account_truncating(),
```

Comment:
Should the destination account here still be updated to Bridgehub's treasury account? [#940 (comment)](https://github.com/Snowfork/snowbridge/pull/940#discussion_r1332961063)

Reply:
Yes, it's treasury underlying https://github.com/Snowfork/cumulus/pull/63/files#r1338607827

Reply:
https://github.com/Snowfork/snowbridge/pull/940/commits/6dfe50247504fbc2c556808496c4a631613d757e for less confusion.